### PR TITLE
feat: generic service notification (CMC-636)

### PR DESCRIPTION
**Pull request by pd-dudkiewicz**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/CMC-639

### Change description ###

Added generic notification to inform solicitors about response to case.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/RespondToClaimCallbackHandler.java [hunk 1]
```diff
@@ -72,6 +76,15 @@ private CallbackResponse setClaimantResponseDeadline(CallbackParams callbackPara
...
         data.put(CLAIMANT_RESPONSE_DEADLINE, claimantResponseDeadLine.atTime(16, 0));
 
+        var response = mapper.convertValue(data.get("respondent1ClaimResponseType"), DefendantResponseType.class);
+        if (response == FULL_DEFENCE) {
+            data.put("businessProcess",
+                     BusinessProcess.builder().activityId("DefendantResponseHandling").status(READY).build());
+        } else {
+            data.put("businessProcess",
+                     BusinessProcess.builder().activityId("CaseHandedOfflineHandling").status(READY).build());
+        }
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/RespondToClaimCallbackHandler.java]:
```suggestion
            data.put("businessProcess",
                     BusinessProcess.builder()
                     .activityId( response == FULL_DEFENCE ? "DefendantResponseHandling" : "CaseHandedOfflineHandling" )
                     .status(READY)
                     .build()
                     );
      
```

File: src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/notification/CaseHandedOfflineDefendantNotificationHandler.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            String.format(REFERENCE_TEMPLATE, caseData.getLegacyCaseReference())
+        );
+
+        return AboutToStartOrSubmitCallbackResponse.builder().build();
+    }
+
+    private Map<String, String> getNotificationProperties(CaseData caseData) {
+        return Map.of(
+            CLAIM_REFERENCE_NUMBER, caseData.getLegacyCaseReference(),
+            SOLICITOR_NAME, caseData.getSolicitorReferences().getRespondentSolicitor1Reference()
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/notification/CaseHandedOfflineDefendantNotificationHandler.java]: `SOLICITOR_NAME` or `SOLICITOR_REFERENCE` ?

File: src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/notification/CaseTransferredToLocalCourtClaimantNotificationHandler.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+    private CallbackResponse notifyClaimantForCaseTransferredToLocalCourt(CallbackParams callbackParams) {
+        CaseData caseData = caseDetailsConverter.toCaseData(callbackParams.getRequest().getCaseDetails());
+
+        notificationService.sendMail(
+            "claimant-solicitor@example.com",
+            notificationsProperties.getSolicitorResponseToCase(),
+            Map.of(
+                CLAIM_REFERENCE_NUMBER, caseData.getLegacyCaseReference(),
+                SOLICITOR_NAME, caseData.getSolicitorReferences().getApplicantSolicitor1Reference()
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/notification/CaseTransferredToLocalCourtClaimantNotificationHandler.java]: `SOLICITOR_NAME` or `SOLICITOR_REFERENCE` ?

File: src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/notification/CaseTransferredToLocalCourtDefendantNotificationHandler.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+    private CallbackResponse notifyDefendantForCaseTransferredToLocalCourt(CallbackParams callbackParams) {
+        CaseData caseData = caseDetailsConverter.toCaseData(callbackParams.getRequest().getCaseDetails());
+
+        notificationService.sendMail(
+            "defendant-solicitor@example.com",
+            notificationsProperties.getSolicitorResponseToCase(),
+            Map.of(
+                CLAIM_REFERENCE_NUMBER, caseData.getLegacyCaseReference(),
+                SOLICITOR_NAME, caseData.getSolicitorReferences().getRespondentSolicitor1Reference()
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/notification/CaseTransferredToLocalCourtDefendantNotificationHandler.java]: same here

File: src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/notification/DefendantResponseClaimantNotificationHandler.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+            getNotificationProperties(caseData),
+            String.format(REFERENCE_TEMPLATE, caseData.getLegacyCaseReference())
+        );
+        return AboutToStartOrSubmitCallbackResponse.builder().build();
+    }
+
+    private Map<String, String> getNotificationProperties(CaseData caseData) {
+        return Map.of(
+            CLAIM_REFERENCE_NUMBER, caseData.getLegacyCaseReference(),
+            SOLICITOR_NAME, caseData.getSolicitorReferences().getApplicantSolicitor1Reference()
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/notification/DefendantResponseClaimantNotificationHandler.java]: same here and other similar places

File: src/main/java/uk/gov/hmcts/reform/unspec/callback/CallbackHandler.java [hunk 6]
```diff
@@ -32,7 +32,7 @@ public boolean isEventAlreadyProcessed(BusinessProcess businessProcess) {
 
     public void register(Map<String, CallbackHandler> handlers) {
         handledEvents().forEach(
-            handledEvent -> handlers.put(handledEvent.getValue(), this));
+            handledEvent -> handlers.put(handledEvent.name(), this));
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/reform/unspec/callback/CallbackHandler.java:35]: 👍

File: src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/notification/CaseHandedOfflineClaimantNotificationHandler.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+            "claimant-solicitor@example.com",
+            notificationsProperties.getSolicitorResponseToCase(),
+            getNotificationProperties(caseData),
+            String.format(REFERENCE_TEMPLATE, caseData.getLegacyCaseReference())
+        );
+
+        return AboutToStartOrSubmitCallbackResponse.builder().build();
+    }
+
+    private Map<String, String> getNotificationProperties(CaseData caseData) {
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/notification/CaseHandedOfflineClaimantNotificationHandler.java]: Should we implement `NotificationData` to be consistent across? I can see that is being currently used to override below method to provide properties.
```
public Map<String, String> addProperties(CaseData caseData) 
```

File: src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/RespondToClaimCallbackHandlerTest.java [hunk 8]
```diff
@@ -103,6 +107,40 @@ void shouldSetClaimantResponseDeadline_whenInvoked() {
...
+            AboutToStartOrSubmitCallbackResponse response = (AboutToStartOrSubmitCallbackResponse) handler
+                .handle(callbackParamsOf(data, CallbackType.ABOUT_TO_SUBMIT));
+
+            assertThat(response.getData()).extracting("businessProcess").extracting("status").isEqualTo(READY);
+            assertThat(response.getData()).extracting("businessProcess").extracting("activityId").isEqualTo(
+                "DefendantResponseHandling");
+            assertThat(response.getData()).extracting("businessProcess").extracting("processInstanceId").isNull();
+        }
+
+        @ParameterizedTest()
```
Comment (sabahirfan) [inline src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/RespondToClaimCallbackHandlerTest.java]:
```suggestion
        @ParameterizedTest
```

File: src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/RespondToDefenceCallbackHandlerTest.java [hunk 9]
```diff
@@ -48,6 +52,58 @@ void shouldReturnNoError_WhenAboutToStartIsInvoked() {
...
+            AboutToStartOrSubmitCallbackResponse response = (AboutToStartOrSubmitCallbackResponse) handler
+                .handle(callbackParamsOf(data, CallbackType.ABOUT_TO_SUBMIT));
+
+            assertThat(response.getData()).extracting("businessProcess").extracting("status").isEqualTo(READY);
+            assertThat(response.getData()).extracting("businessProcess").extracting("activityId").isEqualTo(
+                "CaseTransferredToLocalCourtHandling");
+            assertThat(response.getData()).extracting("businessProcess").extracting("processInstanceId").isNull();
+        }
+
+        @ParameterizedTest()
```
Comment (sabahirfan) [inline src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/RespondToDefenceCallbackHandlerTest.java]:
```suggestion
        @ParameterizedTest
```

File: src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/RespondToDefenceCallbackHandlerTest.java [hunk 10]
```diff
@@ -48,6 +52,58 @@ void shouldReturnNoError_WhenAboutToStartIsInvoked() {
...
+                "applicant1ProceedWithClaim", "Yes"
+            ));
+
+            AboutToStartOrSubmitCallbackResponse response = (AboutToStartOrSubmitCallbackResponse) handler
+                .handle(callbackParamsOf(data, CallbackType.ABOUT_TO_SUBMIT));
+
+            assertThat(response.getData()).doesNotContainKey("businessProcess");
+        }
+
+        @ParameterizedTest()
```
Comment (sabahirfan) [inline src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/RespondToDefenceCallbackHandlerTest.java]:
```suggestion
        @ParameterizedTest
```

File: src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/RespondToClaimCallbackHandler.java (hunk 1 above)
Comment (pd-dudkiewicz) [reply src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/RespondToClaimCallbackHandler.java]: nice, updated 👍

File: src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/notification/CaseHandedOfflineDefendantNotificationHandler.java (hunk 2 above)
Comment (pd-dudkiewicz) [reply src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/notification/CaseHandedOfflineDefendantNotificationHandler.java]: should be reference, updated in all places

File: src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/notification/CaseHandedOfflineClaimantNotificationHandler.java (hunk 7 above)
Comment (pd-dudkiewicz) [reply src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/notification/CaseHandedOfflineClaimantNotificationHandler.java]: updated, wasn't there before rebasing 😄

**Review by sabahirfan — DISMISSED** [review]

**Review by douglasrice — DISMISSED** [review]

**Review by sabahirfan — DISMISSED** [review]

**Review by sabahirfan — DISMISSED** [review]

**Review by douglasrice — DISMISSED** [review]

**Review by douglasrice — APPROVED** [review]

**Review by sabahirfan — APPROVED** [review]

**Pull request merged**